Implemented Clock Class

3saster opened the pull request:
Implemented a UI Clock class into ZScript. _GetRealTime()_ gives seconds since local midnight in order to display the system time. _GetTime()_ gives seconds since Unix epoch, and can be useful as a timing method; for example, to see how long the user has been away upon loading a save.

https://forum.zdoom.org/viewtopic.php?f=15&t=65827

3saster commented:
I've renamed the functions and the structure. It's now called _SystemTime_, and _Now()_ returns the epoch time (I'll leave this as is because it allows a clean way to compare times and define user times independent of time zones). I've added a _Format_ function, which takes an epoch time and formats it the same way as strftime does.

Everything works as intended, with one exception. If an invalid string format is fed to _SystemTime.Format_ (say, "%Q"), on a VS-compiled build at least, it will crash GZDoom as soon as this is line is executed. I'm not really sure how to fix this. Would it be better to manually implement the behaviour of strftime instead?

alexey-lysiuk commented on src/scripting/vmthunks.cpp:
```diff
@@ -3080,16 +3080,40 @@ static int GetRealTime()
 	time(&now);
-	struct tm* timeinfo = localtime(&now);
-	return timeinfo ? timeinfo->tm_sec + timeinfo->tm_min * 60 + timeinfo->tm_hour * 3600 : 0;
+	return now ? now : 0;
```
Error value of `time()` is `(time_t)(-1)`, not zero.

alexey-lysiuk commented on src/scripting/vmthunks.cpp:
```diff
@@ -3080,16 +3080,40 @@ static int GetRealTime()
+static void FormatTime(const FString &timeForm, int timeVal, FString* result)
+{
+	time_t val = timeVal;
+	struct tm* timeinfo = localtime(&val);
```
Returned value needs to be checked for `nullptr`.

alexey-lysiuk commented on src/scripting/vmthunks.cpp:
```diff
@@ -3080,16 +3080,40 @@ static int GetRealTime()
+	time_t val = timeVal;
+	struct tm* timeinfo = localtime(&val);
+	char timeString[1024];
+	strftime(timeString, sizeof(timeString), timeForm, timeinfo);
```
`strftime()` may fail as well potentially leaving `timeString` filled with garbage.

alexey-lysiuk commented on wadsrc/static/zscript/base.zs:
```diff
@@ -418,6 +418,12 @@ struct GameInfoStruct native
 
+struct SystemTime ui
+{
+	native static int Now();
```
It would be nice to have a comment about the value this method returns.

alexey-lysiuk commented on wadsrc/static/zscript/ui/statusbar/alt_hud.zs:
```diff
@@ -770,50 +770,62 @@ class AltHud ui
 
 			if (hud_showtime < 8)
 			{
+				int timeSeconds;
```
No need to separate definition and assignment here.

alexey-lysiuk commented on wadsrc/static/zscript/ui/statusbar/alt_hud.zs:
```diff
@@ -770,50 +770,62 @@ class AltHud ui
+					int millis  = (Level.time % Thinker.TICRATE) * (1000 / Thinker.TICRATE);
+					timeString = SystemTime.Format("%H:%M:%S",timeSeconds);
+					timeString = String.Format("%s.%03i", timeString, millis);
+				}
```
There is no option for system time with milliseconds.

alexey-lysiuk commented on src/scripting/vmthunks.cpp:
```diff
@@ -3080,7 +3080,7 @@ static int GetRealTime()
 	time_t now;
 	time(&now);
-	return now ? now : 0;
+	return now ? now : (time_t)(-1);
```
In this form the comparison is useless. You need to return it as is (so -1 will designate an error), or to check `now` for `(time_t)(-1)` and return zero if it's the case (zero means an error). I would just do the first.

Doom2fan commented:
> 
> 
> IMHO there is no room for this `setdate` thing in the engine. It's a wall of code with no real purpose.
> Modder can pass arbitrary values to `SystemTime.Format()` order to test some special dates/times.

It's not for testing, it's so players can override the date if someone uses it for special events (e.g., some christmas thing) through netevents.

3saster commented:
Yeah, that was Rachel's reasoning, and it makes a lot of sense for that purpose. Perhaps not the cleanest solution, but I don't think users would be keen to type in a unix timestamp instead...

There is one more thing left to do. _strftime_ crashes GZDoom if an invalid format specifier is given. Would it be best to perform parsing to ensure it's valid, or just write _mystrftime_? I'm not entirely clear on whether this is a standard thing, or if microsoft or anyone else made there own additions to it.

alexey-lysiuk commented:
> Yeah, that was Rachel's reasoning, and it makes a lot of sense for that purpose. Perhaps not the cleanest solution, but I don't think users would be keen to type in a unix timestamp instead...

This RotT example didn't convince me. It's only about time display in UI, not about making some decision.
Also, even such a very special case can be tested exclusively on mod's side.

> There is one more thing left to do. _strftime_ crashes GZDoom if an invalid format specifier is given. 

Could you please be more specific? What's an invalid format? Where is it crashes?

> Would it be best to perform parsing to ensure it's valid, or just write _mystrftime_? I'm not entirely clear on whether this is a standard thing, or if microsoft or anyone else made there own additions to it.

`strftime()` is a part of C/C++ standard. If some vendor implements some additions to it, that's the problem of that vendor. Apparently, only specifiers mentioned in standard are supported.

Doom2fan commented:
> This RotT example didn't convince me. It's only about time display in UI, not about making some decision.
> Also, even such a very special case can be tested exclusively on mod's side.

I don't understand why you're so hung up on "testing". Literally *no one* mentioned testing anywhere, the point is so players can trigger effects like Rachael's example at any time they want.

MajorCooke commented:
I vote for keeping it in. You might not see the reason behind it but you can both test events AND enjoy them happening.

Doom2fan commented:
(_mental_ accidentally edited this into my comment instead of posting a comment)
>> I don't understand why you're so hung up on "testing". Literally *no one* mentioned testing anywhere, the point is so players can trigger effects like Rachael's example at any time they want.
>
>What's the point of doing something at a special date if everyone can trigger it anytime he want?
I cannot imaging anything except to test it during development.
>
>The only one kind of code doesn't contain bugs. The one that wasn't written.
>Every line of kinda useful stuff (emphasis on _kinda_) adds complexity and maintenance cost.
And I assure you, it has bugs.
>
>For example, the second call to `localtime` in CCMD's function lacks a check of returned value.
>Do you want to support and fix "yet another very useful feature"? I don't.
>This thing is so niche that it doesn't worth time I spent explaining my thoughts about it.

MajorCooke commented:
Also, Nash's day/night system could definitely benefit from this.

Doom2fan commented:
>> I don't understand why you're so hung up on "testing". Literally *no one* mentioned testing anywhere, the point is so players can trigger effects like Rachael's example at any time they want.
>
>What's the point of doing something at a special date if everyone can trigger it anytime he want?
I cannot imaging anything except to test it during development.

Because not everyone has the free time to play games at whatever time they want. There's also the fact that the ***entire reason*** no such function was added before is because Graf didn't want that. I can assure you that Graf will not allow this function to be added at all without such an override.

>The only one kind of code doesn't contain bugs. The one that wasn't written.
>Every line of kinda useful stuff (emphasis on _kinda_) adds complexity and maintenance cost.
And I assure you, it has bugs.
>
>For example, the second call to `localtime` in CCMD's function lacks a check of returned value.
>Do you want to support and fix "yet another very useful feature"? I don't.
>This thing is so niche that it doesn't worth time I spent explaining my thoughts about it.

What are you even on about? That has literally nothing to do with anything.

MajorCooke commented:
Tyrian is another game that also does something like this.

alexey-lysiuk commented:
Sorry about changing the comment, have no idea why I pressed edit there.

Regarding the topic, `setdate` CCMD still needs to be fixed and reviewed. No offense, but just posting votes won't make it any better.

MajorCooke commented:
What specifically needs to be fixed?

nashmuhandes commented:
> Also, Nash's day/night system could definitely benefit from this.

I noticed I was name-dropped here... no, not really. I have no need/interest to retrieve the system time. Plus as I understand it, this is a UI-only feature, whereas the time cycle in my game is playsim. 

At any rate, I didn't request this, just to clear the air.

madame-rachelle commented:
I did request it though, and I do think such an override would be useful and I *do* believe it is necessary and I agree with what has already been posted here for reasoning (some of which I posted myself). I don't know why it's ballooning into an argument like this - there is really no reason that it should.

If there are technical issues with the override itself, fine, we can work on that and fix it, but in general I do see it as a necessary feature, and I am getting quite disappointed that as of late that there is so much resistance to useful things like this.

At first I did not even want this feature at *all* to start with, and I was reluctant to accept it, but with this override being a compromise for that helped with it. This really is a player-centric issue, in my opinion, *and* it does help modders even so (ie making a special mod for Christmas that does special things on that day, before the actual day occurs, without it being so necessary to fuck with the code as you are suggesting they should).

coelckers commented:
Ugh. One afternoon where I didn't check my mails and this happened.
So here's the final decision, and I don't think it comes as a surprise: 
A date feature that lets people do date-locked features without any chance of override is a no-go. If this goes in the user must also be able to set a date without screwing with the system's global date.

End of discussion.

3saster commented:
So there is one last issue at play here, before this could be successfully merged. The behaviour of _mktime_ is implementation specific, and so the setdate command has different behaviour. On a MSVC compilation, if DST (daylight savings) is currently inactive, but you try to set the date to one where it was active, it will interpret the date with the current DST settings, then apply the ones at the actual date (so that an input of 2PM will display as 3PM). On GCC/Clang, it does this in the other order, so an input of 2PM will always display as 2PM. I hope this issue makes sense. Here is an [example code](https://gist.github.com/3saster/3d341379b8cace0d89730d595b46d235) to demonstrate the different behaviour.

What this amounts to is that what date the clock is set at using the _setdate_ CCMD is dependent on the compiler used. The date itself however, will be set correctly, just not to what the input might have been. This "error" will be at most 1 hour in either direction. How should this be fixed/addressed? I would personally just print the date after it's set to let the person know what the time is, and they can then adjust it themselves. Perhaps there may be a better way to approach this though?

NightFright2k19 commented:
What's the current status on this? Any problems with the requested changes?

NightFright2k19 commented:
Can anybody resolve the remaining file conflict? If that's the only thing that prevents this from being implemented, it could possibly find its way into the next GZD release already.

3saster commented:
It's certainly not the only issue remaining, the file conflict is a trivial fix. The bigger issue is how MSVC has different behaviour of _mktime_, and I'm not clear on how the best way to address it is. Additionally, this is a low priority thing, so it may take some time before anyone actually decides to take a look at it.

MajorCooke commented:
Please fix the merge conflicts so I can merge this into QZDoom for testing.

3saster commented:
@madame-rachelle I've brought it up to date, simplified some things, and generally everything is working now. I even came up with a fix for the DST issue. As such, it's fully functional, but there are a couple of points that should be noted/discussed before this is merged:

- A ui struct called **SystemTime** has been added to base.zs, featuring two functions, **Now** and **FormatTime**. **Now** simply returns the epoch time as an integer, while **FormatTime** formats an epoch time using the strftime specifiers, similar to how console.printf works. I think this works okay for now, but perhaps there is a better interface to use?
- **FormatTime** has an internal function **CheckStrfString**, which essentially checks if the input is legal. The list of legal specifiers is in the _validSingles_ array. Some of these are locale-specific, which I'm not sure whether we want that or not. I previously had the special %E and %O modifiers included, but these are just locale-specific versions of the previous specifiers, and it made **CheckStrfString** a lot larger for something I deem overkill. These can be added back if desired, but I don't really think they're needed personally.
- I've added 2 Console Commands for dealing with the time, _getdate_ and _setdate_. The former simply prints the current date, in HH:MM:SS DD:MM:YYYY syntax. The latter takes the same input and sets the date accordingly. There are various messages for different potential input errors of setdate (including printing out the input syntax), including a message if it was unable to set the date (such as if you went before the dawn of time, AKA Jan 1 1970).
- Internally, setdate works by offsetting the epoch.
- As GetRealTime is redundant, alt_hud.zs was slightly changed, and that (private) function was removed.

madame-rachelle commented:
Thank you. :)
